node/forest_libp2p/src/blocksync/provider.rs

```diff
@@ -0,0 +1,241 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use chain::Error as ChainError;
+use forest_blocks::{Tipset, TipsetKeys};
+use forest_cid::Cid;
+use ipld_blockstore::BlockStore;
+use log::debug;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use super::{
+    BlockSyncRequest, BlockSyncResponse, BlockSyncResponseStatus, CompactedMessages, TipsetBundle,
+};
+
+/// Builds blocksync response out of chain data.
+pub fn make_blocksync_response<DB>(db: Arc<DB>, request: &BlockSyncRequest) -> BlockSyncResponse
+where
+    DB: BlockStore,
+{
+    let mut response_chain: Vec<TipsetBundle> = vec![];
+
+    let mut curr_tipset_cids = request.start.clone();
+
+    loop {
+        let mut tipset_bundle: TipsetBundle = Default::default();
+        let tipset = match chain::tipset_from_keys(db.as_ref(), &TipsetKeys::new(curr_tipset_cids))
+        {
+            Ok(tipset) => tipset,
+            Err(err) => {
+                debug!("Cannot get tipset from keys: {}", err);
+
+                return BlockSyncResponse {
+                    chain: vec![],
+                    status: BlockSyncResponseStatus::InternalError,
+                    message: "Tipset was not found in the database".to_owned(),
+                };
+            }
+        };
+
+        if request.include_blocks() {
+            tipset_bundle.blocks = tipset.blocks().to_vec();
+        }
```

Comment:
Might be worth rearranging this logic so you don't need to clone the headers. I can do this later so no need to worry about it if you don't want

Reply:
moved this logic at the end of the function and used `into_blocks()` instead.